File: Supplementary_webpages/static/gallery_res/Convex Optimization page 276/convex_optimization_276.py
"""
given
A_i ∈ ℝ^(m_i × n)  
b_i ∈ ℝ^m_i 
`x₀` ∈ ℝ^n  

min_(x ∈ ℝ^n) ∑_i ‖A_i x + b_i‖ + (1/2)‖x-`x₀`‖²
"""
import numpy as np
import scipy
import scipy.linalg
from scipy import sparse
from scipy.integrate import quad
from scipy.optimize import minimize


class convex_optimization_276ResultType:
    def __init__( self, ret):
        self.ret = ret


def convex_optimization_276(A, b, x0):
    A = np.asarray(A)
    b = np.asarray(b)
    x0 = np.asarray(x0, dtype=np.float64)

    dim_0 = A.shape[0]
    n = A[0].shape[1]
    assert x0.shape == (n,)

    def target_0(x):
        sum_0 = 0
        for i in range(1, len(b)+1):
            sum_0 += np.linalg.norm(A[i-1] @ x + b[i-1], 2)
        return sum_0 + (1 / 2) * np.power(np.linalg.norm(x - x0, 2), 2)
    ret = minimize(target_0, np.zeros(n)).fun
    return convex_optimization_276ResultType(ret)


def generateRandomData():
    dim_0 = np.random.randint(10)
    n = np.random.randint(10)
    A = []
    b = []
    for i in range(dim_0):
        m_0 = np.random.randint(10)
        A.append(np.random.randn(m_0, n))
        b.append(np.random.randn(m_0))
    x0 = np.random.randn(n)
    return A, b, x0


if __name__ == '__main__':
    A, b, x0 = generateRandomData()
    print("A:", A)
    print("b:", b)
    print("x0:", x0)
    func_value = convex_optimization_276(A, b, x0)
    print("return value: ", func_value.ret)

File: Supplementary_webpages/static/gallery_res/Convex Optimization page 276/convex_optimization_276.cpp
/*
given
A_i ∈ ℝ^(m_i × n)  
b_i ∈ ℝ^m_i 
`x₀` ∈ ℝ^n  

min_(x ∈ ℝ^n) ∑_i ‖A_i x + b_i‖ + (1/2)‖x-`x₀`‖²
*/
#include <Eigen/Core>
#include <Eigen/Dense>
#include <Eigen/Sparse>
#include <iostream>
#include <set>

struct convex_optimization_276ResultType {
    double ret;
    convex_optimization_276ResultType(const double & ret)
    : ret(ret)
    {}
};

convex_optimization_276ResultType convex_optimization_276(
    const std::vector<Eigen::MatrixXd> & A,
    const std::vector<Eigen::VectorXd> & b,
    const Eigen::VectorXd & x₀)
{
    const long dim_0 = A.size();
    const long n = A[0].cols();
    assert( A.size() == dim_0 );
    assert( b.size() == dim_0 );
    assert( x₀.size() == n );

    double ret = ;
    return convex_optimization_276ResultType(ret);
}


void generateRandomData(std::vector<Eigen::MatrixXd> & A,
    std::vector<Eigen::VectorXd> & b,
    Eigen::VectorXd & x₀)
{
    const int dim_0 = rand()%10;
    const int n = rand()%10;
    A.resize(dim_0);
    b.resize(dim_0);
    for(int i=0; i<dim_0; i++){
        int m_1 = rand()%10;
        A[i] = Eigen::MatrixXd::Random(m_1, n);
        b[i] = Eigen::VectorXd::Random(m_1);
    }
    x₀ = Eigen::VectorXd::Random(n);
}


int main(int argc, char *argv[])
{
    srand((int)time(NULL));
    std::vector<Eigen::MatrixXd> A;
    std::vector<Eigen::VectorXd> b;
    Eigen::VectorXd x₀;
    generateRandomData(A, b, x₀);
    convex_optimization_276ResultType func_value = convex_optimization_276(A, b, x₀);
    std::cout<<"return value:\n"<<func_value.ret<<std::endl;
    return 0;
}